sklearn/gp_search.py

```diff
@@ -0,0 +1,421 @@
+"""
+The :mod:`sklearn.gp_search` includes utilities to fine-tune the parameters
+of an estimator through a Gaussian Process model.
+"""
+from __future__ import print_function
+
+# Author: Sebastien Dubois <http://bit.ly/SebastienDubois>,
+# License: BSD 3 clause
+
+import numpy as np
+from scipy.stats import norm
+
+from .gaussian_process import GaussianProcessRegressor
+from .cross_validation import check_cv
+from .cross_validation import _fit_and_score
+from .metrics.scorer import check_scoring
+from .base import is_classifier, clone
+from .grid_search import BaseSearchCV
+
+
+def _sample_candidates(n_candidates, param_bounds, param_is_int):
+    n_parameters = param_is_int.shape[0]
+    candidates = []
+
+    for k in range(n_parameters):
+        if param_is_int[k]:
+            k_sample = np.asarray(
+                np.random.rand(n_candidates)
+                * np.float(param_bounds[k][1]-param_bounds[k][0])
+                + param_bounds[k][0],
+                dtype=np.int32)
+        else:
+            k_sample = np.asarray(
+                np.random.rand(n_candidates)
+                * np.float(param_bounds[k][1]-param_bounds[k][0])
+                + param_bounds[k][0])
+        candidates.append(k_sample)
+
+    candidates = np.asarray(candidates).T
+    return unique_rows(candidates)
+
+
+def compute_expected_improvement(predictions, sigma, y_best):
+    ei_array = np.zeros(predictions.shape[0])
+    for i in range(ei_array.shape[0]):
+        z = (y_best - predictions[i]) / sigma[i]
+        ei_array[i] = sigma[i] * (z * norm.cdf(z) + norm.pdf(z))
+
+    return ei_array
+
+
+def unique_rows(ar):
+    # keep only unique values in the ndarray ar
+    # http://stackoverflow.com/questions/16970982/find-unique-rows-in-numpy-array
+
+    b = np.ascontiguousarray(ar).view(
+        np.dtype((np.void, ar.dtype.itemsize * ar.shape[1]))
+        )
+    _, idx = np.unique(b, return_index=True)
+    idx = np.sort(idx)
+
+    return ar[idx]
+
+
+def is_in_ndarray(item, a):
+    # look for element item in ndarray a
+    # returns True if item is in a, and its index
+
+    k = 0
+    idx_val = np.asarray(range(a.shape[0]))
+    idxk = range(a.shape[0])
+    while(k < a.shape[1]):
+        idxk = (a[idxk, k] == item[k])
+        if np.sum(idxk > 0):
+            k += 1
+            idx_val = idx_val[idxk]
+            idxk = list(idx_val)
+        else:
+            return False, 0
+
+    return True, idx_val[0]
+
+
+class GPSearchCV(BaseSearchCV):
+    """
```

Comment:
General description is missing. It would be helpful to describe the various steps of the algorithm which are referenced in the parameters (n_iter, n_init, n_final_iter).